### PR TITLE
add use-ironic design document

### PR DESCRIPTION
**Pull request by dhellmann**
Start filling in some of the details about using Ironic.

Signed-off-by: Doug Hellmann <dhellmann@redhat.com>

### CODE REVIEW

File: design/use-ironic.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# use-ironic
+
+## Status
+
+One of: implementable
```
Comment (russellb) [inline design/use-ironic.md]: s/One of: //

File: design/use-ironic.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+None
+
+### Implementation Details/Notes/Constraints [optional]
+
+- We want to eliminate the need for a MySQL database or RabbitMQ
+  message bus.
+- We need to include Ironic Inspector as well as the main Ironic API
+  so we can perform hardware introspection.
+- Both services should run inside the same Pod as the
+  baremetal-operator.
```
Comment (russellb) [inline design/use-ironic.md]: Maybe expand on this a bit to explain why that is desirable?

File: design/use-ironic.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+### Version Skew Strategy
+
+Version skew should not apply if the operator and ironic services run
+in the same Pod.
+
+## Drawbacks [optional]
+
+Ideally we would not need a long running service at all, but until we
+have a tool compatible with the Job API we cannot avoid that.
```
Comment (russellb) [inline design/use-ironic.md]: Maybe this is where we could expand on how even though Ironic is a long running service, it's not the primary owner of any given information.  If it was lost, we can reconfigure it from state stored in CRDs.

File: design/use-ironic.md (hunk 1 above)
Comment (dhellmann) [reply design/use-ironic.md]: Done

File: design/use-ironic.md (hunk 2 above)
Comment (dhellmann) [reply design/use-ironic.md]: Done

File: design/use-ironic.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+## Alternatives [optional]
+
+We could run Ironic and Ironic Inspector in their own Pod, outside of
+the one containing the baremetal-operator. This would make local
+development of the operator a little simpler, because developers would
+not need to run the services on their local host. On the other hand,
+it means we have to somehow tell the operator where the two services
+are so it can build a URL to communicate with them. It also means
+something needs to manage the Pod running the services to ensure it is
+present in the cluster when it is needed.
```
Comment (russellb) [inline design/use-ironic.md]: It also simplifies security concerns for the Ironic API

File: design/use-ironic.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+### Implementation Details/Notes/Constraints [optional]
+
+- We want to eliminate the need for a MySQL database or RabbitMQ
+  message bus.
+- We need to include Ironic Inspector as well as the main Ironic API
+  so we can perform hardware introspection.
+- Both services should run inside the same Pod as the
+  baremetal-operator.
+- The Pod containing the services will need host network access to be
+  able to run a PXE service on the provisioning network.
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply design/use-ironic.md]: Done

File: design/use-ironic.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  Ironic.
+
+### Work Items
+
+TBD
+
+### Dependencies
+
+- Ironic image
+- Ironic Inspector image
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply design/use-ironic.md]: Done

File: design/use-ironic.md (hunk 3 above)
Comment (dhellmann) [reply design/use-ironic.md]: Done

File: design/use-ironic.md (hunk 4 above)
Comment (dhellmann) [reply design/use-ironic.md]: Done

File: design/use-ironic.md (hunk 2 above)
Comment (russellb) [reply design/use-ironic.md]: hm, don't see it?  To be clear, I was thinking of explaining somewhere why running within the same pod is a desirable thing.

I do like the additions that explain the MySQL and Rabbit changes!

File: design/use-ironic.md (hunk 3 above)
Comment (russellb) [reply design/use-ironic.md]: Are you referring to the expanded text about MySQL in the details section earlier in the doc?

File: design/use-ironic.md (hunk 2 above)
Comment (russellb) [reply design/use-ironic.md]: Nevermind, I see the security note later in the doc.

File: design/use-ironic.md (hunk 3 above)
Comment (dhellmann) [reply design/use-ironic.md]: Yeah, I thought having it up in the implementation details section made more sense, sorry I should have been explicit in my "done" message. Do you think I should add more here, too?

Comment (russellb) [reply design/use-ironic.md]: All good.  thanks!

**Comment by russellb** [comment]
LGTM, but would you like to squash this first?

**Pull request merged**